onnxruntime/test/providers/qnn/qnn_ep_context_test.cc

```diff
@@ -801,6 +826,296 @@ TEST_F(QnnHTPBackendTests, QnnMultiContextExternal) {
+    std::shared_ptr<Model> ctx_model;
+    auto path_str = ToPathString(ep_ctx_file);
+    ASSERT_STATUS_OK(Model::Load(path_str, ctx_model, nullptr, logger));
+    auto& ctx_graph = ctx_model->MainGraph();
+    GraphViewer graph_viewer(ctx_graph);
+    auto path = std::filesystem::path(path_str);
+
+    for (auto& node : ctx_graph.Nodes()) {
+      if (node.OpType() == "EPContext") {
+        NodeAttrHelper node_helper(node);
+        int64_t is_main_context = node_helper.Get("main_context", static_cast<int64_t>(0));
+        if (1 == is_main_context) {
+          std::string old_qnn_ctx_binary_file_name = node_helper.Get("ep_cache_context", "");
+          auto file_path = path.replace_filename(old_qnn_ctx_binary_file_name);
+          std::remove(file_path.string().c_str());
+          node.ClearAttribute("ep_cache_context");
+          node.AddAttribute("ep_cache_context", last_qnn_ctx_binary_file_name);
+        }
+      }
+    }
+    std::remove(ep_ctx_file.c_str());
+    ASSERT_STATUS_OK(Model::Save(*ctx_model.get(), ToPathString(ep_ctx_file)));
+  }
+}
+
+static void GetModelInputNames(const std::string& model_path,
+                               std::vector<std::string>& input_names,
+                               std::vector<std::string>& output_names,
+                               const Logger& logger) {
+  std::shared_ptr<Model> ctx_model;
+  auto path_str = ToPathString(model_path);
+  ASSERT_STATUS_OK(Model::Load(path_str, ctx_model, nullptr, logger));
+  auto& ctx_graph = ctx_model->MainGraph();
+
+  auto& inputs = ctx_graph.GetInputs();
+  for (auto input : inputs) {
+    input_names.push_back(input->Name());
+  }
+
+  auto& outputs = ctx_graph.GetOutputs();
+  for (auto output : outputs) {
+    output_names.push_back(output->Name());
+  }
+}
+
+// 1. Create 2 QDQ models
+// 2. Initialize 2 Ort sessions which share the same QNN EP from these 2 QDQ models
+// with EpContextEnable = 1, to dump the context binary
+// so, the 2nd context binary contains the graph from the 1st model
+// 3. Change the 1st context model to point to the 2nd context binary file
+// 4. Start 2 ort session from the dumped context model,
+// The 2nd session uses graph from 1st session
+// 5. Run the 2nd session
+TEST_F(QnnHTPBackendTests, QnnContextShareAcrossSessions1) {
+  ProviderOptions provider_options;
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnHtp.dll";
+#else
+  provider_options["backend_path"] = "libQnnHtp.so";
+#endif
+
+  // Create QDQ models
+  std::vector<std::string> onnx_model_paths{"./weight_share1.onnx", "./weight_share2.onnx"};
+  std::vector<std::string> ctx_model_paths;
+  for (auto model_path : onnx_model_paths) {
+    CreateQdqModel(model_path, DefaultLoggingManager().DefaultLogger());
+    EXPECT_TRUE(std::filesystem::exists(model_path.c_str()));
+    ctx_model_paths.push_back(model_path + "_ctx.onnx");
+  }
+
+  DumpModelWithSharedCtx(provider_options, onnx_model_paths);
```

Comment:
Should these provider options also set `enable_htp_weight_sharing`? I didn't see that for these unit tests.

Reply:
To verify the inference code, especially the feature added by this PR that share QNN graphs across Ort sessions, it doesn't matter whether the weight sharing is enabled or not. It just requires multiple QNN graphs in one QNN context, and single Ort session uses part of the graphs.
The consideration here is weight sharing feature is support only by arch >= v73.
We don't want our UT limit by that.